### DISCUSSION TITLE
Expression Mapping Requires Mapping From Destination To Source Why?

### DISCUSSION THREAD

**Discussion by biqas** [AutoMapper/AutoMapper.Extensions.ExpressionMapping#166] [General]
Hi,

If I want to map an expression from source (DTO) to target (Entity), why the mapping definition has to be provided in the direction destination to source instead of source to destination?

Example:
```csharp
public class ExpressionMappingTest
{
    [Fact]
    public void MapExpression()
    {
        var mapper = new MapperConfiguration
        (
            cfg =>
            {
                cfg.AddExpressionMapping();

                //cfg.CreateMap<TenantDto, TenantEntity>(); <-- Why Not This?
                //cfg.CreateMap<MetaDataDto, MetaDataEntity>();

                cfg.CreateMap<TenantEntity, TenantDto>(); <-- Why This
                cfg.CreateMap<MetaDataEntity, MetaDataDto>();
            }
        )
        .CreateMapper();

        Expression<Func<TenantDto, bool>> dtoExpression = x => x.MetaData.CreatedBy != "";
        var mapped = mapper.MapExpression<Expression<Func<TenantEntity, bool>>>(dtoExpression);
    }

    public record TenantDto
    {
        public Guid UID { get; init; }
        public string Name { get; set; }
        public MetaDataDto MetaData { get; init; }
    }

    public record MetaDataDto
    {
        public DateTimeOffset CreatedOn { get; init; }
        public string CreatedBy { get; init; }
    }

    public record UserDto
    {
        public Guid UID { get; init; }
        public Guid TenantUID { get; init; }
        public TenantDto? Tenant { get; init; }
        public MetaDataDto MetaData { get; init; }
    }

    public record TenantEntity
    {
        public string Name { get; set; }
        public Guid UID { get; init; }
        public MetaDataEntity MetaData { get; init; }
    }

    public record MetaDataEntity
    {
        public DateTimeOffset CreatedOn { get; init; }
        public string CreatedBy { get; init; }
    }
}
```

this works, but why the mapping has to be provided in this direction "`cfg.CreateMap<TenantEntity, TenantDto>();`" it is an expression without assignments so why can't AutoMapper deal with this: "`cfg.CreateMap<TenantDto, TenantEntity>();`"

**Comment by BlaiseD** [comment]
There are [comments](https://github.com/AutoMapper/AutoMapper.Extensions.ExpressionMapping/blob/826680034397364df9b37d5b50ad8220834a41b2/src/AutoMapper.Extensions.ExpressionMapping/XpressionMapperVisitor.cs#L670) in the source describing why.   Running some of the tests will make it clear.

**Reply by biqas** [reply]
I would agree to this, if there would be assignments, but in this case, it is only mapping/translating one-way. I see from the comments why it is necessary, but this does not consider if you just want to pass predicates as dto's and wants them mapped to entity predicates, means no reverse assignment will ever happen.
